### ISSUE TITLE
Concurrent rendering with multiple mbgl.Map instances

### ISSUE THREAD

**Issue by petrsloup** [mapbox/mapbox-gl-native#4215]
When using the node package (3.0.2) is it possible to have multiple instances of `mbgl.Map` and using them to render more images simultaneously (since `mbgl.Map#render` does not allow this)? It seems to be possible, but is it safe?

More specifically, I'm trying to run this inside a docker container with xvfb emulating an X framebuffer.

Running the `npm test` in docker with xvfb:
```bash
docker run -it --rm node:5.7.1 /bin/bash
apt-get update && apt-get install -y xvfb
...
xvfb-run --server-args="-screen 0 1024x768x24" npm test
```
This works ok with the current tests.

But if I modify some of the test (for example https://github.com/mapbox/mapbox-gl-native/blob/master/platform/node/test/js/map.test.js#L245) to:
```javascript
        t.test('returns an image', function(t) {
            var map = new mbgl.Map(options);
            var map2 = new mbgl.Map(options);
            map.load(style);
            map2.load(style);
            map.render({}, function(err, pixels) {
                t.error(err);
                map.release();
                t.ok(pixels);
                t.ok(pixels instanceof Buffer);
                t.equal(pixels.length, 512 * 512 * 4)
            });
            map2.render({}, function(err, pixels) {
                t.error(err);
                map2.release();
                t.ok(pixels);
                t.ok(pixels instanceof Buffer);
                t.equal(pixels.length, 512 * 512 * 4)
                t.end();
            });
        });
```
the test sometimes passes, sometimes segfaults and sometimes prints out various shader errors.
For example:
```
terminate called after throwing an instance of 'mbgl::util::ShaderException'
  what():  Program plain failed to link: error: uniform `gl_LightSource' declared as type `gl_LightSourceParameters[8]' and type `gl_LightSourceParameters[8]'
```

```
terminate called after throwing an instance of 'mbgl::util::ShaderException'
  what():  Program plain failed to link: error: uniform `gl_BackLightProduct' declared as type `gl_LightProducts[8]' and type `gl_LightProducts[8]'
```

```
terminate called after throwing an instance of 'mbgl::util::ShaderException'
  what():  Program plain failed to link: error: uniform `gl_CurrentAttribVertMESA' declared as type `vec4[33]' and type `vec4[33]'
```

(This makes it somewhat similar to errors reported in https://github.com/cutting-room-floor/node-mapbox-gl-native/issues/18.)

So, is it okay to create more instances of `mbgl.Map` to render concurrently or is this not supported?

**Comment by jfirebaugh** [comment]
Yes, multiple instances of `mbgl.Map` for concurrent rendering is supported. The errors you posted don't look familiar to me -- mapbox-gl-native doesn't use uniforms named `gl_LightSource` and `gl_BackLightProduct`. Are you using custom shaders?

Also, note in your test code, `map2.render` doesn't necessarily complete after `map.render`. You want to arrange for `t.end` to be called only after both renders complete.

**Comment by jfirebaugh** [comment]
I believe there are also some concurrency bugs in older versions of MESA -- you may need to use a more recent version that what your distribution provides.

**Comment by petrsloup** [comment]
@jfirebaugh Thanks, good to hear this is supported!
Turns out this is actually bug in mesa3d -- upgrading debian from `stable` to `testing` seems to have fixed this, so there is probably some concurrency bug in mesa3d that got fixed somewhere between `10.3.2` and `11.1.2` (just FYI in case someone runs into this issue).

Thanks for the advice!
